src/curve_server.cpp

```diff
@@ -410,62 +404,52 @@ int zmq::curve_server_t::process_initiate (msg_t *msg_)
             //  reply already, but removing this has some strange side-effect
             //  (probably because the pipe's in_active flag is true until a read
             //  is attempted)
-            rc = receive_and_process_zap_reply ();
-            if (rc == -1)
-                goto exit;
+            if (-1 == receive_and_process_zap_reply ())
+                return -1;
         } else if (!options.zap_enforce_domain) {
             //  This supports the Stonehouse pattern (encryption without
             //  authentication) in legacy mode (domain set but no handler).
             state = sending_ready;
         } else {
             session->get_socket ()->event_handshake_failed_no_detail (
               session->get_endpoint (), EFAULT);
-            rc = -1;
-            goto exit;
+            return -1;
         }
     } else {
         //  This supports the Stonehouse pattern (encryption without authentication).
         state = sending_ready;
     }
 
-    rc = parse_metadata (initiate_plaintext + crypto_box_ZEROBYTES + 128,
-                         clen - crypto_box_ZEROBYTES - 128);
-
-exit:
-    free (initiate_plaintext);
-    free (initiate_box);
-    return rc;
+    return parse_metadata (&initiate_plaintext[crypto_box_ZEROBYTES + 128],
+                           clen - crypto_box_ZEROBYTES - 128);
 }
 
 int zmq::curve_server_t::produce_ready (msg_t *msg_)
 {
     const size_t metadata_length = basic_properties_len ();
     uint8_t ready_nonce[crypto_box_NONCEBYTES];
 
-    uint8_t *ready_plaintext =
-      static_cast<uint8_t *> (malloc (crypto_box_ZEROBYTES + metadata_length));
-    alloc_assert (ready_plaintext);
+    std::vector<uint8_t> ready_plaintext (crypto_box_ZEROBYTES
+                                          + metadata_length);
 
     //  Create Box [metadata](S'->C')
-    memset (ready_plaintext, 0, crypto_box_ZEROBYTES);
-    uint8_t *ptr = ready_plaintext + crypto_box_ZEROBYTES;
+    std::fill (ready_plaintext.begin (),
+               ready_plaintext.begin () + crypto_box_ZEROBYTES, 0);
+    uint8_t *ptr = &ready_plaintext[crypto_box_ZEROBYTES];
 
     ptr += add_basic_properties (ptr, metadata_length);
-    const size_t mlen = ptr - ready_plaintext;
+    const size_t mlen = ptr - &ready_plaintext[0];
```

Comment:
is pointer arithmetic OK to use with std::vectors?

Reply:
Yes, unfortunately the syntax is somewhat awkward since std:: vector::data only exists from C++11. However, replacing memcpy by std::copy would be preferable. I planned to do this in another step.

Reply:
great, thanks

Reply:
We should actually use libsodium's secure memory zones for secrets: https://libsodium.gitbook.io/doc/memory_management

Reply:
It should be possible to implement a custom allocator that uses sodium_malloc/sodium_free, that can be used with std::vector.

Reply:
tweetnacl doesn't have those functions though, but I can implement a fallback for the allocator.

Reply:
yeah tweetnacl shouldn't be used in production, so it's fine if it falls back to standard malloc

Reply:
Opened #3576 for this. I already started implementing this in https://github.com/sigiesec/libzmq/tree/use-libsodium-based-allocator

Reply:
thanks!